### PR TITLE
Module to create Route53 zone

### PR DESCRIPTION
**Pull request by sai-caylent**
https://strongmind.atlassian.net/browse/EXCAY2-44?atlOrigin=eyJpIjoiNjlmODVkYjhkNzY2NGI4NGFiODg0NWNkMjVmNTI3Y2EiLCJwIjoiaiJ9

### CODE REVIEW

File: deployment/src/strongmind_deployment/acm.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    domain: str = "prod.aws2.strongmind.com"
+
+
+class AcmCertificate(pulumi.ComponentResource):
+    def __init__(
+        self, name: str, args: AcmCertificateArgs, opts: pulumi.ResourceOptions = None
+    ):
+        super().__init__("pkg:index:AcmCertificate", name, None, opts)
+        child_opts = pulumi.ResourceOptions(parent=self)
+        if args.stack != "prod":
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/acm.py:22]: we should make sure this is centralized - the prod logic.

File: deployment/src/strongmind_deployment/acm.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    zone_id: str
+    tags: dict = None
+    domain: str = "prod.aws2.strongmind.com"
+
+
+class AcmCertificate(pulumi.ComponentResource):
+    def __init__(
+        self, name: str, args: AcmCertificateArgs, opts: pulumi.ResourceOptions = None
+    ):
+        super().__init__("pkg:index:AcmCertificate", name, None, opts)
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/acm.py:20]: there's a precedence for package naming in the other components,  we should follow that.

File: deployment/src/strongmind_deployment/acm.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        child_opts = pulumi.ResourceOptions(parent=self)
+        if args.stack != "prod":
+            name = f"{args.stack}-{name}"
+        domain = args.domain
+        full_name = f"*.{domain}"
+
+        self.cert = acm.Certificate(
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/acm.py:27]: lets follow the pattern I've established for the module design so they look similar,   we can drop the old style from the previous engagement.

File: deployment/src/strongmind_deployment/acm.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import re
+from dataclasses import dataclass
+from pulumi import Output, ResourceOptions
+from pulumi_aws import acm, route53
+import pulumi
+
+
+@dataclass
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/acm.py:8]: this dataclass has introduced issues.  I don't think Mark was a big fan of it either, so we can remove that.  

Pydantic is an option but might need to test it out a bit first.  I had some issues with pydantic.

File: deployment/src/strongmind_deployment/route53_zone.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def __str__(self):
+        return self.value
+
+
+class Route53Zone(pulumi.ComponentResource):
+    zone: aws.route53.Zone
+    domain_name: str
+
+    def __init__(self, name, args: Route53ZoneArgs, opts=None):
+        super().__init__("custom:Route53:Zone", name, {}, opts)
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/route53_zone.py:33]: resource naming again.  
There is precedence: 
https://github.com/StrongMind/public-reusable-workflows/blob/0f02af92aa42641b7271fd66a51dfa35836d99ff/deployment/src/strongmind_deployment/container.py#L37

**Review by brettswift — CHANGES_REQUESTED** [review]

File: deployment/src/strongmind_deployment/acm.py (hunk 4 above)
Comment (markng) [reply deployment/src/strongmind_deployment/acm.py:8]: I'm actually good with dataclasses, pydantic is like dataclasses++, as it has validations (which might be really good for making sure that teams don't send non-useful elements to the components, and improving their user experience when they do so).

File: deployment/src/strongmind_deployment/acm.py (hunk 2 above)
Comment (markng) [reply deployment/src/strongmind_deployment/acm.py:20]: Agreed.

**Review by markng — CHANGES_REQUESTED** [review]

File: deployment/src/strongmind_deployment/acm.py (hunk 2 above)
Comment (sai-caylent) [reply deployment/src/strongmind_deployment/acm.py]: done

File: deployment/src/strongmind_deployment/acm.py (hunk 4 above)
Comment (sai-caylent) [reply deployment/src/strongmind_deployment/acm.py]: replaced with pydantic

File: deployment/src/strongmind_deployment/route53_zone.py (hunk 5 above)
Comment (sai-caylent) [reply deployment/src/strongmind_deployment/route53_zone.py]: followed this

File: deployment/src/strongmind_deployment/acm.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            return acm.CertificateValidation(
+                "cert_validation",
+                certificate_arn=self.cert.arn,
+                validation_record_fqdns=[resource_record_value],
+                opts=ResourceOptions(
+                    parent=self, depends_on=[self.validation_dns_record_value]
+                ),
+            )
+
+    def get_name(self, stack: str, name: str):
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/acm.py:107]: we shouldn't have code like this in libraries.   it shouldn't know about it's stack.

File: deployment/src/strongmind_deployment/route53_zone.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+class Route53ZoneArgs(BaseModel):
+    domain_name: str
+    vpc_id: Optional[str] = None
+    vpc_region: Optional[str] = None
+    comment: Optional[str] = "Managed by Pulumi"
+    force_destroy: Optional[bool] = False
+    delegation_set_id: Optional[str] = None
+    tags: Optional[Dict[str, str]] = None
+
+
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/route53_zone.py:26]: I probably named this Enum in the old code.   They aren't outputs though.   So it's not an accurate name.

Presumably these are used in the consumer code to set exports in shared_infrastructure.

maybe a better name is `R53ZoneExportNames`.

File: deployment/src/strongmind_deployment/acm.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    domain: str = "aws2.strongmind.com"
+
+
+class AcmCertificate(pulumi.ComponentResource):
+    def __init__(
+        self, name: str, args: AcmCertificateArgs, opts: pulumi.ResourceOptions = None
+    ):
+        super().__init__(
+            "strongmind:global_build:commons:acmcertificate", name, None, opts
+        )
```
Comment (brettswift) [inline deployment/src/strongmind_deployment/acm.py:30]: to keep the look and feel similar,   lets set primitive property types in the constructor directly, but use a `create_resources()` function that will create the certificate resources here.

**Review by brettswift — CHANGES_REQUESTED** [review]

**Review by brettswift — APPROVED** [review]

**Review by markng — APPROVED** [review]

**Pull request merged**